cli/commands/catalog/module/doc.go

```diff
@@ -0,0 +1,307 @@
+package module
+
+import (
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/gruntwork-io/go-commons/errors"
+)
+
+const (
+	mdExt   = ".md"
+	adocExt = ".adoc"
+
+	docTitle docDataKey = iota
+	docDescription
+	docContent
+
+	tagH1Block docTagName = iota
+	tagH2Block
+)
+
+var (
+	// `strings.EqualFold` is used (case insensitive) while comparing
+	docFiles = []string{"README.md", "README.adoc"}
+
+	frontmatterKeys = map[string]docDataKey{
+		"name":        docTitle,
+		"description": docDescription,
+	}
+)
+
+type docDataKey byte
+type docTagName byte
+
+type DocRegs []*regexp.Regexp
+
+func (regs DocRegs) Replace(str string) string {
+	for _, reg := range regs {
+		str = reg.ReplaceAllString(str, "$1")
+	}
+	return str
+}
+
+type Doc struct {
+	rawContent string
+	fileExt    string
+
+	tagCache     map[docDataKey]string
+	tagRegs      map[docTagName]*regexp.Regexp
+	tagStripRegs DocRegs
+
+	frontmatterCache map[docDataKey]string
+	frontmatterReg   *regexp.Regexp
+}
+
+func newDoc(rawContent, fileExt string) *Doc {
+	doc := &Doc{
+		rawContent: rawContent,
+		fileExt:    fileExt,
+
+		tagRegs:        make(map[docTagName]*regexp.Regexp),
+		frontmatterReg: regexp.MustCompile(`(?i)^[\s\n]*<!-- frontmatter[\s\n]*([\S\s]*?)[\s\n]*-->`),
+	}
+
+	switch fileExt {
+	case mdExt:
+		doc.tagRegs[tagH1Block] = regexp.MustCompile(`(?:^|\n)\#{1}\s([\S\s]+?)(?:[\r\n]+\#|[\r\n]*$)`)
+		doc.tagRegs[tagH2Block] = regexp.MustCompile(`(?:^|\n)\#{2}\s([\S\s]+?)(?:[\r\n]+\#|[\r\n]*$)`)
+		doc.tagStripRegs = DocRegs{
```

Comment:
Will these work if something wraps over multiple lines? E.g.,:

```md
[Here is a
link that wraps over multiple lines](https://www.gruntwork.io)
```

Also, what about nested expressions?

```md
> [Here is **bold text** and `code` inside a link](https://www.gruntwork.io) all of which is in a quote block
```

I believe that's all valid Markdown...

Reply:
It's worth noting there are Markdown parsers... Maybe one of those would be a better way to handle this?

Reply:
> Will these work if something wraps over multiple lines? E.g.,:
> 
> ```
> [Here is a
> link that wraps over multiple lines](https://www.gruntwork.io)
> ```

Checked, this is the result:

```
Here is a
link that wraps over multiple
```

> 
> Also, what about nested expressions?
> 
> ```
> > [Here is **bold text** and `code` inside a link](https://www.gruntwork.io) all of which is in a quote block
> ```

Checked, this is the result:
```
Here is bold text and code inside a link all of which is in a quote block
```

So it works pretty good.

Reply:
> It's worth noting there are Markdown parsers... Maybe one of those would be a better way to handle this?

The most important is to convert `md/adoc` into a plain text. I tried to find ready-made solutions before implementation but found nothing for Golang.